### ISSUE TITLE
Move JS tests to GitHub Actions

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#2366]
## Feature Description

As part of our initiative to move our CI jobs to GitHub Actions, this issue is for moving our JS tests. This corresponds to the job that runs on Travis by the same name.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A new GitHub Actions workflow definition should be created for "JS Tests"
* The workflow should contain a job for running all the same JS tests that we run on Travis
* The workflow and job should fail under the same conditions as they do today when running on Travis
* All JS tests, JS-only variables, and script steps should be removed from `.travis.yml`
* Previous runs of the workflow should be cancelled automatically by newer runs using our `cancel.yml` workflow by [adding the workflow file name of the workflow to the list provided to `workflow_id`](https://github.com/google/site-kit-wp/blob/07d95eddab386c3c5b9f8d1dad1f123d7cb33dae/.github/workflows/cancel.yml#L23) 
* The workflow should be a required check (set in repo settings)

## Implementation Brief

* Add a `js-tests.yml` file to our `.github/workflows/` folder.
* Following the format of the storybook tests, add our existing "JS tests": `bundlesize` and `jest`
  * This would be a good opportunity to better name the `bundlesize` test: it's currently run using `npm test`, which is awkward especially for new developers, thinking `npm test` is what they expect: the unit tests for the project. Let's instead have `npm test` run `npm run test:js` and move bundlesize to `npm run test:bundlesize`
* Merge #2535.
* Set the Repo Settings to require the new workflow named "JS Tests". (@felixarntz will need to handle this)

### Test Coverage

* None needed.

### Visual Regression Changes

* No changes.

## QA Brief

* The PR (#2535) should pass and use GitHub Actions to run the new check named "JS Tests":

<img width="1084" alt="Screenshot 2020-12-17 at 00 51 06" src="https://user-images.githubusercontent.com/90871/102423898-f4bb6980-4001-11eb-95dd-89827738353a.png">

## Changelog entry

* N/A

**Comment by aaemnnosttv** [comment]
@tofumatt IB looks good for the most part.

> This would be a good opportunity to better name the `bundlesize` test: it's currently run using `npm test`, which is awkward especially for new developers, thinking `npm test` is what they expect: the unit tests for the project. Let's instead rename `npm test` [...]

Agreed 👍 This is also consistent with the command for running PHPUnit tests (`composer test`).

> Let's instead rename `npm test` to be `npm run test:js:watch` (common for JS projects to have their watch tests as the default command—it's a good developer experience) and move bundlesize to `npm run test:bundlesize`

I'm not sure we should make the `test` command watch by default. While it may be common for JS projects to do this, it's rather inconsistent with our other commands as none of them watch without the added `:watch` part in the name. I think it might be worth removing this at some point since I find it easier to just add `--watch` at the end since I'm usually not running the full suite anyways, but that's beyond the scope here.

---

I missed this when writing the ACs here, but if it's not too much effort, I think it might be worth extracting Bundlesize to it's own check as described here:
https://github.com/siddharthkp/bundlesize#build-status-for-github

We could make this a follow-up issue (it doesn't have to block/change the work here), but I think that would be a nice enhancement and keep this job running faster as well since it wouldn't require an actual build.

PS. I updated the PR # in the IB since it referenced a non-existent issue 😉 

---

TL;DR – this is basically good to go, just the one reservation about the command for `test` including watch 😄

**Comment by tofumatt** [comment]
Unfortunately the bundlesize check requires the `dist/` files be built and checked in the repo, from what I can see. There isn't a provision for building them first, so I don't think that'll work.

We could separate out each test to be its own job, but given it takes `npm ci` about a minute to run even with a cache, best to keep things all in this one job for now I think.

I've updated the IB to use `npm run test:js` though, fair points! 👍🏻

**Comment by aaemnnosttv** [comment]
> Unfortunately the bundlesize check requires the `dist/` files be built and checked in the repo, from what I can see. There isn't a provision for building them first, so I don't think that'll work.

Not sure where you're seeing that but looking at some of the projects that use it listed on the repo, you can see how it's implemented: https://github.com/redux-saga/redux-saga/blob/9d2164946f402e594a0dfe453c6d20fb6f14858f/.github/workflows/nodejs.yml#L33-L36

It's not a big deal, but would be a nice touch to include; we can always revisit.

IB ✅

**Comment by tofumatt** [comment]
Oh, well then, maybe it runs `npm run build` automatically? Sorry, that step is just totally opaque to me and not mentioned anywhere in the docs that I could find.

But I'll give it a go in the PR; if it works let's just ship it here rather than making a whole new issue for it. Agreed: I'd like it as well! 👍🏻

**Comment by tofumatt** [comment]
Ah, I see. Looks like @felixarntz is the best person to add that in as I can't authorise the Google GitHub account to have access to bundlesize and the token needs to be added to GitHub secrets for this repo.

The action is set up to run bundlesize though, using this secret name: `BUNDLESIZE_GITHUB_TOKEN`. Adding the value obtained from https://github.com/login/oauth/authorize?scope=repo%3Astatus&client_id=6756cb03a8d6528aca5a should do the trick!

**Comment by eugene-manuilov** [comment]
QA :warning: Overall it looks good, however, the workflow file contains the `ready_for_review` activity type which doesn't make sense there. @tofumatt could you please remove it? Also, why do we need the following step? Does it test something?

https://github.com/google/site-kit-wp/blob/6f6ea93b7971159bc506b8d6decaf0af1c4c59f2/.github/workflows/js-tests.yml#L41-L42

**Comment by tofumatt** [comment]
Oh, the `ready_for_review` I think came from a suggestion from @felixarntz, but I can't recall now. If it doesn't make sense happy to remove—the `npm run build:test` command is there so we can run `bundlesize` on the built files.

**Comment by tofumatt** [comment]
Also covered by #2549; seemed wasteful to make a PR for each 😄

**Comment by asvinb** [comment]
Checked here:
https://github.com/google/site-kit-wp/actions/runs/429338349

![image](https://user-images.githubusercontent.com/31818/102994927-79921000-4539-11eb-8788-ba4520098456.png)

JS tests passing

QA: ✅
